src/common/selftest/selftest_firstlayer.cpp

```diff
@@ -0,0 +1,342 @@
+/**
+ * @file selftest_firstlayer.cpp
+ */
+
+#include "selftest_firstlayer.hpp"
+#include "wizard_config.hpp"
+#include "i_selftest.hpp"
+#include "filament_sensor_api.hpp"
+#include "filament.hpp"
+#include "../../Marlin/src/gcode/queue.h"
+#include "../../lib/Marlin/Marlin/src/gcode/lcd/M73_PE.h"
+#include "../../lib/Marlin/Marlin/src/gcode/gcode.h"
+#include "../../Marlin/src/module/probe.h"
+#include "../../Marlin/src/module/temperature.h"
+#include "../../marlin_stubs/G26.hpp"
+#include "M70X.hpp"
+
+#include <array>
+
+using namespace selftest;
+
+CSelftestPart_FirstLayer::CSelftestPart_FirstLayer(IPartHandler &state_machine, const FirstLayerConfig_t &config,
+    SelftestFirstLayer_t &result)
+    : rStateMachine(state_machine)
+    , rConfig(config)
+    , rResult(result)
+    , filament_known_but_unsensed(false)
+    , state_selected_by_user(StateSelectedByUser::Calib)
+    , log(1000) {
+    rStateMachine.SetTimeToShowResult(0);
+}
+
+LoopResult CSelftestPart_FirstLayer::stateStart() {
+    LogInfo("%s Started", rConfig.partname);
+    SelftestInstance().log_printf("%s Started\n", rConfig.partname);
+
+    return LoopResult::RunNext;
+}
+
+LoopResult CSelftestPart_FirstLayer::stateAskFilamentInit() {
+    uint8_t filament = 0;
+    filament |= Filaments::CurrentIndex() != filament_t::NONE ? FKNOWN : 0;
+    filament |= FSensors_instance().GetPrinter() == fsensor_t::NoFilament ? F_NOTSENSED : 0;
+
+    filament_known_but_unsensed = false;
+    // 4 posible states
+    // 0 !FKNOWN !F_NOTSENSED
+    // 1 FKNOWN !F_NOTSENSED
+    // 2 !FKNOWN F_NOTSENSED
+    // 3 FKNOWN F_NOTSENSED
+    switch (filament) {
+    case FKNOWN: //known and not "unsensed" - do not allow load
+        filament_known_but_unsensed = true;
+        rStateMachine.SetFsmPhase(PhasesSelftest::FirstLayer_filament_known_and_not_unsensed);
+        rResult.preselect_response = Response::Next;
+        break;
+    case FKNOWN | F_NOTSENSED: //allow load, prepick UNLOAD, force ask preheat
+        rStateMachine.SetFsmPhase(PhasesSelftest::FirstLayer_filament_not_known_or_unsensed);
+        rResult.preselect_response = Response::Unload;
+        break;
+    case F_NOTSENSED: //allow load, prepick LOAD, force ask preheat
+    case 0:           //filament is not known but is sensed == most likely same as F_NOTSENSED, but user inserted filament into sensor
+    default:
+        rStateMachine.SetFsmPhase(PhasesSelftest::FirstLayer_filament_not_known_or_unsensed);
+        rResult.preselect_response = Response::Load;
+        break;
+    }
+    LogInfo("%s user asked about filament");
+    return LoopResult::RunNext;
+}
```

Comment:
Function is not documented. There are some comments inside function body but those are hard to understand. Code is hard to understand, especially I don't like using negative flag F_NOTSENSED and unnamed 0.

Reply:
I would like if the method would get the header with the documentation of what it does or what is its intended purpose. I would like if the comments inside the function body would be removed or reduced to absolute minimum.

Reply:
not sensed means filament sensor is enabled and working  properly while does not detect any filament
negation of this flag means all other states

will do it better way